### PR TITLE
refactor shard selection

### PR DESCRIPTION
**Pull request by Sovianum**
Minor refactoring before adding explicit replica weights support

https://joom.myjetbrains.com/youtrack/issue/JMINFRA-231/Dobavit-topology-awareness-v-redispipe

### CODE REVIEW

**Review by g7r — CHANGES_REQUESTED** [review]

File: rediscluster/mapping.go [hunk 1]
```diff
@@ -247,90 +247,122 @@ var rr, rs = func() ([32]uint32, [32]uint32) {
...
 	}
-	if conn == nil {
-		c.ForceReloading()
-		return nil, c.err(ErrNoAliveConnection).WithProperty(redis.EKSlot, slot).WithProperty(EKPolicy, policy)
+
+	return nil
+}
+
+func (c *Cluster) weightsForPolicySlaves(policy ReplicaPolicyEnum, shard *shard) []uint32 {
+	var ws [32]uint32
```
[the remark this answers is not in the document]
Comment (Sovianum) [reply rediscluster/mapping.go]: Changed ws type to []uint32

File: rediscluster/mapping.go [hunk 2]
```diff
@@ -247,90 +247,122 @@ var rr, rs = func() ([32]uint32, [32]uint32) {
...
+		if policy == PreferSlaves {
+			ws = rs
+		}
+	} else {
+		for i := range shard.pingWeights {
+			ws[i] = atomic.LoadUint32(&shard.pingWeights[i])
+		}
 	}
-	return conn, nil
+	return ws[:len(shard.addr)]
```
[the remark this answers is not in the document]
Comment (Sovianum) [reply rediscluster/mapping.go]: Fixed

File: rediscluster/mapping.go [hunk 3]
```diff
@@ -247,90 +247,122 @@ var rr, rs = func() ([32]uint32, [32]uint32) {
...
+
+	return nil
+}
+
+func (c *Cluster) weightsForPolicySlaves(policy ReplicaPolicyEnum, shard *shard) []uint32 {
+	var ws [32]uint32
+	if atomic.LoadUint32(&c.latencyAwareness) == disabled {
+		ws = rr
+		if policy == PreferSlaves {
+			ws = rs
```
[the remark this answers is not in the document]
Comment (Sovianum) [reply rediscluster/mapping.go]: Changed to slicing

File: rediscluster/mapping.go [hunk 4]
```diff
@@ -247,90 +247,122 @@ var rr, rs = func() ([32]uint32, [32]uint32) {
...
+	}
+	return node.getConn(c.opts.ConnHostPolicy, preferConnected, seen)
+}
+
+func (c *Cluster) connForPolicySlaves(policy ReplicaPolicyEnum, seen []*redisconn.Connection, shard *shard, cfg *clusterConfig) *redisconn.Connection {
+	weights := c.weightsForPolicySlaves(policy, shard)
+
+	health := atomic.LoadUint32(&shard.good) // load health information
+	healthWeight := c.getHealthWeight(weights, health)
+	off := c.opts.RoundRobinSeed.Current()
```
[the remark this answers is not in the document]
Comment (Sovianum) [reply rediscluster/mapping.go:292]: As particularly this part of refactoring is not necessary for the main PR https://github.com/joomcode/redispipe/pull/28 removed  `connForPolicySlavesByWeights` function at all and inlined its contents to `connForPolicySlaves`

File: rediscluster/mapping.go [hunk 5]
```diff
@@ -247,90 +247,122 @@ var rr, rs = func() ([32]uint32, [32]uint32) {
...
+
+	// First, we try already established connections.
+	// If no one found, then connections thar are connecting at the moment are tried.
+	for _, needState := range []int{needConnected, mayBeConnected} {
+		if conn := c.connForPolicySlavesByWeights(health, healthWeight, weights, &off, needState, seen, shard, cfg); conn != nil { // off can be modified from inside the function
```
[the remark this answers is not in the document]
Comment (Sovianum) [reply rediscluster/mapping.go]: Comment removed with the function itself

**Review by g7r — APPROVED** [review]

**Pull request merged**